dali/pipeline/operators/paste/paste.cu

```diff
@@ -0,0 +1,179 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/pipeline/operators/paste/paste.h"
+
+#include <utility>
+#include <vector>
+#include <algorithm>
+
+namespace dali {
+
+namespace {
+
+__global__
+__launch_bounds__(512, 1)
+void BatchedPaste(
+    const int N,
+    const int C,
+    const uint8* const __restrict__ fill_value,
+    const uint8* const * const __restrict__ in_batch,
+    uint8* const* const __restrict__ out_batch,
+    const int* const __restrict__ in_out_dims_paste_yx) {
+  const int n = blockIdx.x;
+
+  constexpr int nWaves = 16;
+  constexpr int blockSize = 512;
+  constexpr int nThreadsPerWave = blockSize / nWaves;
+  constexpr int MAX_C = 1024;
+
+  __shared__ uint8 rgb[MAX_C];
+  __shared__ int jump[MAX_C];
+  for (int i = threadIdx.x; i < C; ++i) {
+    rgb[i] = fill_value[i % C];
+    jump[i] = (i + nThreadsPerWave) % C;
+  }
+
+  const int offset = n*6;
+  const int in_H = in_out_dims_paste_yx[offset];
+  const int in_W = in_out_dims_paste_yx[offset + 1];
+  const int out_H = in_out_dims_paste_yx[offset + 2];
+  const int out_W = in_out_dims_paste_yx[offset + 3];
+  const int paste_y = in_out_dims_paste_yx[offset + 4];
+  const int paste_x = in_out_dims_paste_yx[offset + 5];
+
+  const uint8* const input_ptr = in_batch[n];
+  uint8 * const output_ptr = out_batch[n];
+
+  __syncthreads();
+
+  const int myWave = threadIdx.x / nThreadsPerWave;
+  const int myId = threadIdx.x % nThreadsPerWave;
+
+  const int paste_x_stride = paste_x * C;
+  const int in_stride = in_W * C;
+  const int startC = myId % C;
+
+  for (int h = myWave; h < out_H; h += nWaves) {
+    const int H = h * out_W * C;
+    const int in_h = h - paste_y;
+    const bool h_in_range = in_h >= 0 && in_h < in_H;
+    if (h_in_range) {
+      int c = startC;
+      for (int i = myId; i < paste_x * C; i += nThreadsPerWave) {
+        const int out_idx = H + i;
+        output_ptr[out_idx] = rgb[c];
+        c = jump[c];
+      }
+      const int current_in_stride = in_h*in_stride - paste_x_stride;
+      for (int i = myId + paste_x * C; i < (paste_x + in_W) * C; i += nThreadsPerWave) {
+        const int out_idx = H + i;
+        const int in_idx = current_in_stride + i;
+
+        output_ptr[out_idx] = input_ptr[in_idx];
+      }
+      c = startC;
+      for (int i = myId + (paste_x + in_W) * C; i < out_W * C; i += nThreadsPerWave) {
+        const int out_idx = H + i;
+        output_ptr[out_idx] = rgb[c];
+        c = jump[c];
+      }
+    } else {
+      int c = startC;
+      for (int i = myId; i < out_W * C; i += nThreadsPerWave) {
+        const int out_idx = H + i;
+        output_ptr[out_idx] = rgb[c];
+        c = jump[c];
+      }
+    }
+  }
+}
+
+}  // namespace
+
+
+template<>
+void Paste<GPUBackend>::RunHelper(DeviceWorkspace *ws) {
+  BatchedPaste<<<batch_size_, 512, 0, ws->stream()>>>(
```

Comment:
Could we extract this value to one common place instead of having it hard coded here and there?

Reply:
You mean 512? We could do define, yes

Reply:
Yes, 512.